### PR TITLE
feat: allow numbers in partition names and namespaces

### PR DESCRIPTION
**Pull request by mr-cal**
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

Allow hyphens in partitions, partition namespaces, and namespaced partitions.

Unblocks [LP#2069783](https://bugs.launchpad.net/snapcraft/+bug/2069783)
(CRAFT-3043)

### CODE REVIEW

**Review by mattculler — APPROVED** [review]
Nice.  This will be a good one for me to look back at when I add support for hyphens in the partition names, though that change will also have to check for environment variable collisions.

**Comment by mr-cal** [comment]
> Nice. This will be a good one for me to look back at when I add support for hyphens in the partition names, though that change will also have to check for environment variable collisions.

I made an issue and referenced it in the code so we don't foget about it: #789, https://github.com/canonical/craft-parts/pull/790/files#diff-419889a2643383d0b091576acef698b735583eccd40f18ea2eb10298c7fd524bR133

**Comment by mattculler** [comment]
> > Nice. This will be a good one for me to look back at when I add support for hyphens in the partition names, though that change will also have to check for environment variable collisions.
> 
> I made an issue and referenced it in the code so we don't foget about it: #789, https://github.com/canonical/craft-parts/pull/790/files#diff-419889a2643383d0b091576acef698b735583eccd40f18ea2eb10298c7fd524bR133

You are the man!

File: craft_parts/lifecycle_manager.py [hunk 1]
```diff
@@ -76,8 +76,11 @@ class LifecycleManager:
         matching this name.
     :param project_vars: A dictionary containing project variables.
     :param partitions: A list of partitions to use when the partitions feature is
-        enabled. The first partition must be "default" and all partitions must be
-        lowercase alphabetical.
+        enabled. The first partition must be "default" and partitions must contain only
+        lowercase alphabetical and numbers. Partitions may have an optional namespace
+        prefix separated by a forward slash. Namespaces must contain only lowercase
+        alphabetical and numbers and namespaced partitions must only contain lowercase
+        alphabetical, numbers, and hyphens.
```
Comment (lengau) [inline craft_parts/lifecycle_manager.py:83]:
```suggestion
        enabled. The first partition must be "default". Partitions must contain only
        lowercase alphanumeric characters. Partitions may have an optional namespace
        prefix separated by a forward slash, which must contain only lowercase
        alphanumeric characters. Namespaced partitions must only contain lowercase
        alphanumeric characters and hyphens.
```

File: craft_parts/utils/partition_utils.py [hunk 2]
```diff
@@ -22,26 +22,28 @@
...
+)
 
 
 def validate_partition_names(partitions: Optional[Sequence[str]]) -> None:
     """Validate the partition feature set.
 
     If the partition feature is enabled, then:
       - the first partition must be "default"
-      - each partition must contain only lowercase alphabetical characters
+      - each partition must contain only lowercase alphabetical characters and numbers
```
Comment (lengau) [inline craft_parts/utils/partition_utils.py:38]:
```suggestion
      - each partition must contain only lowercase alphanumeric characters
```

File: craft_parts/utils/partition_utils.py [hunk 3]
```diff
@@ -22,26 +22,28 @@
...
       - partitions are unique
 
     Namespaced partitions can also be validated in addition to regular (or
     'non-namespaced') partitions. The format is `<namespace>/<partition>`.
 
     Namespaced partitions have the following naming convention:
-      - the namespace must contain only lowercase alphabetical characters
-      - the partition must contain only lowercase alphabetical characters and hyphens
+      - the namespace must contain only lowercase alphabetical characters and numbers
+      - the partition must contain only lowercase alphabetical characters, numbers, and hyphens
```
Comment (lengau) [inline craft_parts/utils/partition_utils.py:46]:
```suggestion
      - the namespace must contain only lowercase alphanumeric characters
      - the partition must contain only lowercase alphanumeric characters and hyphens
```

**Review by lengau — APPROVED** [review]

**Pull request merged**